src/nvim/api/buffer.c

```diff
@@ -86,6 +86,12 @@ void buffer_del_line(Buffer buffer, Integer index, Error *err)
   FUNC_ATTR_DEFERRED
 {
   Array array = ARRAY_DICT_INIT;
+  buf_T *buf = find_buffer_by_handle(buffer, err);
+
+  if(index >= buf->b_ml.ml_line_count) {
+    return;
```

Comment:
The rationale was that the user's goal of removing the lines has been accomplished either way. Although the lines that were actually removed may be different from the intended ones, this non-atomicity is also a problem with an in-bounds set_line which does not return an error.